### PR TITLE
Add the DS18B20 1-Wire temperature sensor

### PR DESCRIPTION
**Pull request by rnestler**
This adds a basic 1-Wire driver and implements the DS18B20 temperature sensor on top of it.

### CODE REVIEW

File: platformio/lib/OneWire/OneWire.cpp [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+    } else {
+        _pin.output();
+        _pin = 0;
+        wait_us(60);
+        _pin.input();
+    }
+}
+
+// output byte d (least sig bit first).
+void OneWire::write_byte(unsigned char d)
```
Comment (ubruhin) [inline platformio/lib/OneWire/OneWire.cpp]: In the header file the parameter `d` is declared as `uint8_t`, not `unsigned char`.

File: platformio/src/main.cpp [hunk 2]
```diff
@@ -16,6 +17,23 @@ int send_command(I2C& i2c, uint8_t address, uint16_t command) {
     return i2c.write(address, cmd, sizeof(cmd));
 }
 
+void ds18b20_send_command(OneWire& one_wire, uint8_t command) {
```
Comment (ubruhin) [inline platformio/src/main.cpp]: Why are the `ds18b20` functions in `main.cpp`? I think a separate class would make more sense.

File: platformio/src/main.cpp [hunk 3]
```diff
@@ -16,6 +17,23 @@ int send_command(I2C& i2c, uint8_t address, uint16_t command) {
     return i2c.write(address, cmd, sizeof(cmd));
 }
 
+void ds18b20_send_command(OneWire& one_wire, uint8_t command) {
+        one_wire.reset();
+        one_wire.write_byte(0xCC); // skip ROM
```
Comment (ubruhin) [inline platformio/src/main.cpp]: The magic number could be replaced by a more expressive enum or constant.

File: platformio/lib/OneWire/OneWire.cpp [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+{
+    for (uint8_t n=8; n!=0; n--)
```
Comment (ubruhin) [inline platformio/lib/OneWire/OneWire.cpp]: Why counting down? Counting up would be more readable in my opinion...

File: platformio/src/main.cpp (hunk 2 above)
Comment (rnestler) [reply platformio/src/main.cpp]: Because this was late written almost throw away code :wink:

Comment (dbrgn) [reply platformio/src/main.cpp]: @rnestler are you saying that you don't always write ✨PERFECT✨ code???!!!

File: platformio/lib/OneWire/OneWire.cpp (hunk 4 above)
Comment (rnestler) [reply platformio/lib/OneWire/OneWire.cpp]: Hmm... I think I had a different approach in at some place, where I used the `n`. But it doesn't make any sense this way.

File: platformio/lib/OneWire/OneWire.cpp (hunk 1 above)
Comment (rnestler) [reply platformio/lib/OneWire/OneWire.cpp]: It's the same, but yeah inconsistency sucks.

Comment (rnestler) [reply platformio/lib/OneWire/OneWire.cpp]: Done

File: platformio/lib/OneWire/OneWire.cpp (hunk 4 above)
Comment (rnestler) [reply platformio/lib/OneWire/OneWire.cpp]: Done

File: platformio/src/main.cpp (hunk 2 above)
Comment (rnestler) [reply platformio/src/main.cpp]: Done

File: platformio/src/main.cpp (hunk 3 above)
Comment (rnestler) [reply platformio/src/main.cpp]: Done

**Comment by rnestler** [comment]
@ubruhin @dbrgn Good like that?

File: platformio/lib/DS18B20/DS18B20.cpp [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+{
+    _one_wire.reset();
+    _one_wire.write_byte(static_cast<uint8_t>(Command::SkipROM));
+    _one_wire.write_byte(static_cast<uint8_t>(command));
+}
+
+bool DS18B20::wait_for_completion()
+{
+    for(int n=0;n<100; ++n) {
+        uint8_t byte = _one_wire.read_byte();
```
Comment (ubruhin) [inline platformio/lib/DS18B20/DS18B20.cpp]: The local variable is not really required:
```cpp
if (_one_wire.read_byte()) {
    return false;
}
```

File: platformio/lib/DS18B20/DS18B20.cpp [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+        }
+        wait_ms(10);
+    }
+    return true;
+}
+
+float DS18B20::read_temperature()
+{
+    send_command(Command::ReadScratchpad);
+    int16_t temperature = 0;
```
Comment (ubruhin) [inline platformio/lib/DS18B20/DS18B20.cpp]: These 5 lines:
```
int16_t temperature = 0;
for(int n=0;n<2; ++n) {
    uint8_t byte = _one_wire.read_byte();
    temperature |= byte << (8*n);
}
```
could be simplified to:
```
int16_t temperature = _one_wire.read_byte();
temperature |= _one_wire.read_byte() << 8;
```

**Review by ubruhin — APPROVED** [review]
> @ubruhin @dbrgn Good like that?

Much better! :smiley: 

I still have added some comments to show how I would simplify the code... ;)

File: platformio/lib/DS18B20/DS18B20.cpp (hunk 6 above)
Comment (rnestler) [reply platformio/lib/DS18B20/DS18B20.cpp]: hmm manual loop unrolling :wink:

Comment (rnestler) [reply platformio/lib/DS18B20/DS18B20.cpp]: Done

File: platformio/lib/DS18B20/DS18B20.cpp (hunk 5 above)
Comment (rnestler) [reply platformio/lib/DS18B20/DS18B20.cpp]: Done

**Review by ubruhin — APPROVED** [review]
:+1:

**Pull request merged**